design/one-pager-k8s-native-providers.md

```diff
@@ -0,0 +1,287 @@
+not a foreign concept in Crossplane. Existing claim reconcilers create a
+Kubernetes object to represent the external resource. However, Kubernetes-native
+providers differ in that the external resource is actually a Kubernetes object
+itself. This concept is also not new in Crossplane, as [complex workloads]
+operate in this manner. `KubernetesApplication` instances are first scheduled to
+a Kubernetes cluster by the scheduler controller, which causes them to pass the
+predicates of the application controller. It uses the template(s) in the
+`KubernetesApplication` instance to create one or more
+`KubernetesApplicationResource` objects, which are then picked up by the
+resource scheduler. It is responsible for actually creating the
+Kubernetes-native (e.g. `Deployment`, `ConfigMap`, etc.) objects in the target
+cluster, then managing them as a managed reconciler does for an external
+resource. In this scenario, the Kubernetes-native objects are the external
+resources.
+
+## Goals
+
+* Create a common pattern that can be used across infrastructure stacks (and
+  potentially the existing complex workload controllers) that utilize the
+  Kubernetes API as the infrastructure provider API
+* Demonstrate the usage of this pattern with Rook
+
+## Differences from `KubernetesApplication`
+
+While the `KubernetesApplication` pattern detailed above provides a starting
+point for this design, a key difference is that `KubernetesApplication` does not
+adhere to the separation of concern structure of other Crossplane resources
+because Kubernetes resources are not modeled in a `claim <-> portable class <->
+non-portable class <-> managed <-> external` workflow. Kubernetes-native
+infrastructure resources, on the other hand, *should* be modeled with the same
+separation of concern that other cloud-provider resources expose.
+
+## Workflow Design
+
+The general design from a user experience perspective should differ very little
+from utilizing any other infrastructure provider stack. Developers should be
+able to request the creation of an abstract resource type via a claim, and
+cluster operators should be able to define the classes of service available to
+satisfy that claim kind.
+
+As an example, Rook provides the option to provision [Minio] object storage in a
+Kubernetes cluster. Assuming you have installed Rook and the corresponding Minio
+operator into a cluster, you should be able to dynamically provision a Minio
+object storage cluster into the Kubernetes cluster from your Crossplane control
+cluster. The experience would look as follows:
+
+1. Create a Kubernetes `Provider` for target cluster:
+
+```yaml
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: demo-provider-rook
+  namespace: rook-infra-dev
+type: Opaque
+data:
+  endpoint: MY_K8S_ENDPOINT
+  username: MY_K8S_USERNAME
+  password: MY_K8S_PASSWORD
+  clusterCA: MY_K8S_CLUSTER_CA
+  clientCert: MY_K8S_CLIENT_CERT
+  clientKey: MY_K8S_CLIENT_KEY
+  token: MY_K8S_TOKEN
```

Comment:
> This pattern almost makes me wonder whether Kubernetes workloads (KubernetesApplication and friends) should be targeting a Kubernetes Provider rather than a KubernetesCluster directly. Doing so would effectively solve the "BYO cluster" scenario for Kubernetes workloads without having to support "importing" a Kubernetes cluster as a literal KubernetesCluster resource claim.

This isn't strictly relevant to this design, but I'd like to capture it somewhere. Not sure if it belongs in the future considerations of this document or as a separate issue. 🤔

Reply:
I think it should be called out in the future considerations, and also tracked by #497. I will add a comment to this effect on the issue.

Reply:
> I'm a little rusty on how our existing KubernetesCluster connection secret format looks

I second this. Today, if I have a k8s cluster that I want to import or use as `Provider`,  I have to be _really_ careful about what goes to where. By _really_, I mean I couldn't do it. I believe using the kubeconfig file directly as value of 1 field in the secret makes manual interaction easier. Our go k8s clients have support for that format and it's easy to use the client to do the parsing when fine-grained field extraction is needed.

> This pattern almost makes me wonder whether Kubernetes workloads (KubernetesApplication and friends) should be targeting a Kubernetes Provider rather than a KubernetesCluster directly.

I see that this makes things easier but it sounds to me that we'd be bending the definition of `Provider` too much to adopt this usage. One can manually refer to the provider's secret during k8s cluster import if they wanted to do that.

Reply:
Given that under this proposal we'll be expecting some users to write Kubernetes `Provider` secrets by hand, I agree with @muvaf that "create a secret whose key encodes your kubeconfig file" feels simpler than "create a secret whose keys match this Crossplane-specific format".

That said, our current format is a little closer to the Kubernetes service account format, and I'm hesitant to predicate this work on changing how KubernetesCluster connection secrets are formed. I'm also unsure whether it would be a good idea to align on kubeconfig based secrets given that moves us further from compatibility with the format used by service accounts.

Perhaps we should:

1. Solve this with documentation to begin with - i.e. make sure we thoroughly document the expected Kubernetes provider secret format.
1. Raise an issue to track potential improvements?

What do you think @hasheddan?

Reply:
Sounds like a good plan! I will open an issue to track now, and I will make sure that this is reflected in documentation when Kubernetes `Provider` is implemented.

Reply:
#861 created to track potential improvements